nemo/collections/llm/gpt/data/utils.py

```diff
@@ -799,60 +799,93 @@ def _preprocess(
+    Input source as conversations format is also supported and is converted to HF format, however mask and type are
+    ignored. Mask will apply to all non-assistant output tokens.
+        {"conversations": [
+            {"from": "User","value":"<text>"},
+            {"from": "Assistant","value":"<text>", "mask": "User", "system": "<text>", "type": "TEXT_TO_VALUE"}
+        ]}
     """
-    Given a conversation list (source) this function applies the following transformations:
-    1. Convert JSONL conversation format to chat message format.
-    2. Tokenize the chat message by applying the chat template.
-    3. Calculate the mask for the assistant tokens.
+    if not hasattr(tokenizer.tokenizer, "apply_chat_template"):
+        raise ValueError("Cannot apply chat template with tokenizer that is not a HuggingFace AutoTokenizer")
+
+    tools = None
+    if isinstance(source, dict):
+        tools = source.get("tools") or tool_schemas
+
+        if source.get("conversations"):
+            # Detect if Nemo {"conversations": [{"from": "User/Assistant", "value": ""}]}
+            # Convert to HuggingFace chat template [{"role": "stystem/user/assistant", "content": ""}]
+            chat = [{"role": convo["from"].lower(), "content": convo["value"]} for convo in source["conversations"]]
+            if source.get("system"):
+                chat.insert(0, {"role": "system", "content": source["system"]})
+
+        elif source.get("messages"):
+            # HuggingFace chat template {"messages": [{"role": "system/user/assistant", "content": ""}]}
+            chat = source.get("messages")
+    else:
+        chat = source
 
-    Args:
-        hf_chat_dict: dict, the chat message dict from HuggingFace tokenizer.
-        tokenizer: TokenizerSpec, the tokenizer object.
-    """
-    # Use HuggingFace tokenizer to apply the chat template.
     template_has_generation_kwd = (
         '{% generation %}' in tokenizer.tokenizer.chat_template
     )  # assistant mask only works if chat template has generation keyword
-    tokens = tokenizer.tokenizer.apply_chat_template(
-        hf_chat_dict['messages'],
+
+    tokenized_chat = tokenizer.tokenizer.apply_chat_template(
+        chat,
+        tools=tools,
+        tokenize=True,
         return_dict=True,
         return_assistant_tokens_mask=template_has_generation_kwd,
-        return_tensors='pt',
     )
-    input_ids = tokens['input_ids'][0]
+
+    # Choose the last conversation as answer other history are context by finding the last masked token
+    # which indicates end of context and beginning of answer
+    input_ids = tokenized_chat.get("input_ids")
     if template_has_generation_kwd:
-        mask = torch.tensor(tokens['assistant_masks']).to(bool)
+        mask = torch.tensor(tokenized_chat['assistant_masks']).to(bool)
     else:
         mask = torch.ones_like(input_ids)
-    return dict(input_ids=input_ids, mask=mask)
+    if tokenizer.eos_id:
+        input_ids += [tokenizer.eos_id]
```

Comment:
does HF tokenizer not automatically add EOS?

Reply:
Not unless you explicitly add it in the template.

Reply:
Brings up a good point though, should be checking input_ids[-1] to verify that it isn't already eos_id

Reply:
Added the appropriate check here.